File: quicksort/quicksort.rb
# frozen_string_literal: true
require 'pry'
# Implementation of inplace Quicksort
class Quicksort

  def public_sort(a)
    a = a.shuffle
    sort(a, 0, a.length - 1)
    a
  end

  private

  def sort(a, lo, hi)
    return if hi <= lo

    j = partition(a, lo, hi)
    sort(a, lo, j - 1)
    sort(a, j + 1, hi)
  end

  def partition(a, lo, hi)
    i = lo
    j = hi + 1
    v = a[i]

    while true
      while a[i += 1] < v
        break if i == hi
      end
      while v < a[j -= 1]
        break if j == lo
      end
      break if i >= j

      a[i], a[j] = a[j], a[i]
    end
    a[lo], a[j] = a[j], a[lo]
    j
  end
end

test = ["Q", "U", "I", "C", "K", "S", "O", "R", "T", "E", "X", "A", "M", "P", "L", "E"]
binding.pry
quicksort = Quicksort.new
example = quicksort.publi_sort(test)
p example

File: quicksort/quicksort.js
class QuickSort {
  sort(a) {
    a = this.shuffle(a); //Shuffling is needed for perfomance guarantee.
    this.privateSort(a,0, a.length - 1)
    return a;
  }

  privateSort(a, lo, hi) {
    if(hi <= lo) { return }
    let j = this.partition(a, lo, hi);
    this.privateSort(a, lo, j - 1)
    this.privateSort(a, j + 1, hi)
  }

  partition(a, lo, hi) {
    let i = lo;
    let j = hi + 1;
    let v = a[lo];

    while (true) {
      while (a[++i] < v) {
        if (i >= hi) { break; };
      }
      while (v < a[--j]) {
        if (j <= lo) { break; };
      }
      if (i >= j) { break; }
      this.swap(a, i, j);
    }
    this.swap(a, lo, j);
    return j;
  }

  swap(a, firstEl, secondEl) {
    let temp = a[firstEl];
    a[firstEl] = a[secondEl];
    a[secondEl] = temp;
  }

  // Knuth-Fisher-Yates shuffle algorithm
  shuffle(arr) {
    for (var i = arr.length - 1; i > 0; i--) {
      const swapIndex = Math.floor(Math.random() * (i + 1))
      const currentEl = arr[i]
      const elToSwap  = arr[swapIndex]

      arr[i] = elToSwap;
      arr[swapIndex] = currentEl;
    }
    return arr;
  }
}

let sample = ["Q", "U", "I", "C", "K", "S", "O", "R", "T", "E", "X", "A", "M", "P", "L", "E"];
const obj = new QuickSort();
sample = obj.sort(sample);
console.log(sample)
